### PR TITLE
Fixes chemfoam reactions

### PR DESCRIPTION
**Pull request by lewcc**
## What Does This PR Do
Fixes issues introduced in https://github.com/ParadiseSS13/Paradise/pull/25436.

This fixes chemfoam causing runtimes when trying to transfer reagents from foam to ghosts, and not transferring reagents to carbons/living mobs at all.

## Why It's Good For The Game
Foam should work, and it shouldn't explode with runtimes. Sorry!

## Testing
image

## Changelog
:cl:
fix: Chemfoam will now properly transfer reagents to carbon mobs.
/:cl:

### CODE REVIEW

**Review by Spaghetti-bit — APPROVED** [review]

**Review by Sirryan2002 — APPROVED** [review]

**Review by DGamerL — APPROVED** [review]

File: code/game/objects/effects/effect_system/effects_foam.dm [hunk 1]
```diff
@@ -45,12 +45,13 @@
...
-			var/amount = M.reagents.get_reagent_amount(reagent_id)
-			if(amount < max_reagent_filling)
-				M.reagents.add_reagent(reagent_id, min(round(amount / 2), 15))
+	if(iscarbon(A) && !QDELETED(A))
+		var/mob/living/carbon/M = A
+		for(var/datum/reagent/R in reagents.reagent_list)
+			var/amount = M.reagents?.get_reagent_amount(R.id)
+			var/foam_content_amount = reagents.get_reagent_amount(R.id)
+			if(amount < max_reagent_filling && M.reagents)
+				M.reagents.add_reagent(R.id, min(round(foam_content_amount / 2), 15))
```
Comment (DGamerL) [inline code/game/objects/effects/effect_system/effects_foam.dm:54]:
```suggestion
			if(amount < max_reagent_filling)
				M.reagents?.add_reagent(R.id, min(round(foam_content_amount / 2), 15))
```

File: code/game/objects/effects/effect_system/effects_foam.dm [hunk 2]
```diff
@@ -45,12 +45,13 @@
...
 
-	if(ismob(A) && !QDELETED(A))
-		var/mob/M = A
-		for(var/reagent_id in reagents.reagent_list)
-			var/amount = M.reagents.get_reagent_amount(reagent_id)
-			if(amount < max_reagent_filling)
-				M.reagents.add_reagent(reagent_id, min(round(amount / 2), 15))
+	if(iscarbon(A) && !QDELETED(A))
+		var/mob/living/carbon/M = A
+		for(var/datum/reagent/R in reagents.reagent_list)
```
Comment (DGamerL) [inline code/game/objects/effects/effect_system/effects_foam.dm:50]:
```suggestion
		for(var/datum/reagent/R as anything in reagents.reagent_list)
```

File: code/game/objects/effects/effect_system/effects_foam.dm [hunk 3]
```diff
@@ -45,12 +45,13 @@
...
 		reagents.reaction(A, REAGENT_TOUCH, fraction)
 
-	if(ismob(A) && !QDELETED(A))
-		var/mob/M = A
-		for(var/reagent_id in reagents.reagent_list)
-			var/amount = M.reagents.get_reagent_amount(reagent_id)
-			if(amount < max_reagent_filling)
-				M.reagents.add_reagent(reagent_id, min(round(amount / 2), 15))
+	if(iscarbon(A) && !QDELETED(A))
+		var/mob/living/carbon/M = A
```
Comment (DGamerL) [inline code/game/objects/effects/effect_system/effects_foam.dm:49]: Could you rename this variable?

**Review by DGamerL — COMMENTED** [review]
Sorry, just noticed this!

**Review by DGamerL — APPROVED** [review]

**Pull request merged**